Feature/55 data domain registration login

JudjinGM opened the pull request:
Implemented data layer for login only, to know is it okay

JudjinGM committed:
Token data layer implementation

JudjinGM committed:
Remote data layer implementation

JudjinGM committed:
Merge branch 'dev' into feature/55-data_domain_registration_login

JudjinGM committed:
Authentication data layer implementation, only loginByEmail.

JudjinGM committed:
added comment and moved packages

artwist-polyakov commented on app/src/main/java/app/cashadvisor/authorization/data/model/response/ResetPasswordByEmailResponse.kt:
```diff
@@ -0,0 +1,11 @@
+import kotlinx.serialization.SerialName
+import kotlinx.serialization.Serializable
+
+@Serializable
```
почему бы не использовать Parcelable, оно работает сильно быстрее  (подключается через `id("kotlin-parcelize"`) 

![](https://miro.medium.com/v2/resize:fit:1200/format:webp/1*d4iAcVhmfIrbGR4c0yqCvw.png)

JudjinGM replied:
Забыл про Parcelable, заменю 👌

avanisimov replied:
в общем случае по производительности неважно serializable или Parcelable
но если не лень, то лучше Parcelable

avanisimov replied:
Но вообще, если над Response(DTO сетевого слоя) висит Parcelable или Serializable - это знак, что что-то пошло не туда

avanisimov replied:
Саша всех запутал)
Serializable - это от kotlinx.serializtion, для json
так что изначально всё правильно написано.

artwist-polyakov replied:
@avanisimov  Так подожди, нам же приходит от сервера ответ в виде некоторой json структуры.
чтобы ее превратить в дата-класс нам вообще ничего не требуется.

То если мы хотим этот класc потом во что-то сериализовать: например в кеш или в базу  или какой нибудь бандл или еще куда, то надо сериализовать. 
Для этого мы можем аннотировать Serializable или Parcelable.

Я вынес логику, что если хочется аннотировать что-то сериалазеблом — то нужно аннотировать паселабл, потому что последний просто быстрее работает и вообще лучше.

Можешь объяснить, где я запутался?

avanisimov replied:
этот респонс помечен аннотацией kotlinx.serialization.Serializable
она нужна чтобы json от сервера в котлин класс превратить. при этом, этот респонс нигде никуда сохранять не надо, так что Parcelable вообще не понадобится

avanisimov commented on app/src/main/java/app/cashadvisor/authorization/data/network/AuthApiService.kt:
```diff
@@ -0,0 +1,66 @@
+        @Body authByServiceRequest: AuthByServiceRequest
+    ): LoginByServiceResponse
+
+    @Headers("Content-Type: application/json")
```
Content-Type: application/json
это называется Content Negotiation. Правильнее будет это один раз настроить в рамках Retrofit/OkhttpClient и не укзаывать над каждым отдельным методом
в целом это и происходит в NetworkModule 
 `.addConverterFactory(Json.asConverterFactory("application/json".toMediaType()))`

JudjinGM replied:
Если не указывать этот хедер бэк присылает ошибку: 400 "Invalid Content-Type, expected application/json: ", хотя я все указал в NetworkModule. Может ли это быть проблема бэка?

avanisimov replied:
вряд ли в беке, раз с аннотацией работает
тут что-то с настройкой в NetworkModule

avanisimov commented on app/src/main/java/app/cashadvisor/authorization/data/repositoryImpl/AuthRepositoryImpl.kt:
```diff
@@ -0,0 +1,43 @@
+        return try {
+            val data = authRemoteDataSource.loginByEmail(authInputData)
+            // какая-то логика по обновлению токена
+            tokenLocalDataSource.saveAccessToken(data.accessToken)
+            tokenLocalDataSource.saveAccessToken(data.refreshToken)
```
как мы в начале обсуждали
хранение токенов=наличие аккаунта/сессии в памяти приложения
процесс аутентификации/авторизации и хранение/наличие сессии это разные фичи
процесс auth в конце может создать аккаунт в приложении, но сама репа не хранит данные о пользователе

Можем это проговорить голосом на встрече, но общая идея такая
репа авторизации можно только дернуть запрос и может быть сохранить промежуточные данные между шагами процесса авторизации
но хранит токены(аккаунт) должен другой репозиторий

JudjinGM replied:
Понял, поправлю!

avanisimov commented on app/src/main/java/app/cashadvisor/common/data/network/ErrorInterceptor.kt:
```diff
@@ -0,0 +1,51 @@
+        if (!networkConnectionProvider.isConnected()) {
+            throw NetworkException.NoInternetConnection()
+        }
+        val response = with(chain) { proceed(request()) }
```
тут можно обойтись без with
не стоит усложнять код, без особой причины
chain.proceed(request()) - вполне достаточная конструкция

JudjinGM replied:
Понял, поправлю!

JudjinGM replied:
тут получается chain.proceed(chain.request()

avanisimov replied:
да, правильно)

avanisimov commented on app/src/main/java/app/cashadvisor/common/domain/ExceptionToErrorMapperBase.kt:
```diff
@@ -0,0 +1,45 @@
+import app.cashadvisor.common.utill.extensions.logNetworkError
+import java.net.ConnectException
+
+abstract class ExceptionToErrorMapperBase {
```
правильней будет назвать этот класс BaseExceptionToErrorMapper

JudjinGM replied:
Понял, поправлю!

avanisimov commented on app/src/main/java/app/cashadvisor/authorization/domain/ExceptionToErrorMapperAuth.kt:
```diff
@@ -0,0 +1,31 @@
+import app.cashadvisor.common.utill.exceptions.AuthException
+import app.cashadvisor.common.utill.extensions.logDebugError
+
+class ExceptionToErrorMapperAuth : ExceptionToErrorMapperBase() {
```
правильнее будет назвать этот класс AuthExceptionToErrorMapper

JudjinGM replied:
Понял, поправлю!

avanisimov commented on app/src/main/java/app/cashadvisor/common/data/network/ErrorCodeMapper.kt:
```diff
@@ -0,0 +1,26 @@
+class ErrorCodeMapper @Inject constructor() {
+
+    fun getException(message: String, responseCode: Int, statusCode: Int): IOException? {
+        var result: IOException? = when (responseCode) {
+            400 -> NetworkException.BadRequest(message, statusCode)
+            401 -> NetworkException.Unauthorized(message, statusCode)
+            500 -> NetworkException.InternalServerError(message, statusCode)
+            else -> {
+                null
+            }
+        }
+
+        //с сервера должен приходить уникальный код ошибки (statusCode), а не повтор responseCode
+        when (statusCode) {
+            401 -> result = AuthException.InvalidEmailOrPassword(message, statusCode)
+        }
+        return result
+    }
```
тут немного не понял логику
сначала мы смотрим на responseCode
а потом, вне зависимости от вычислений в первом when, мы смотрим на statusCode

JudjinGM replied:
я здесь, пока нет точной инфы как будет организованы ошибки, временно сделал логику, что изначально мы смотрим респонс код, если же есть более точные данные в statusCode, то переписываем его. Позже буду править в зависимости от реализации на беке.

avanisimov replied:
тогда логичнее сначала смотреть на statusCode, тк он более приоритетный

JudjinGM replied:
На данный момент в интерцепторе проверяю только response code. После того, как бэк определится с ошибками, внесу изменения в маппер ошибок приходящий как зависимость в датасорс.

avanisimov commented on app/src/main/java/app/cashadvisor/common/data/network/ErrorInterceptor.kt:
```diff
@@ -0,0 +1,51 @@
+        var error: IOException? = null
+        try {
+            val errorResponse =
+                Json.decodeFromString<ErrorResponse>(responseBody.string())
```
тут Json используется в виде глобальной статической переменной
это опция хороша только для helloworld приложения
по правильному надо настроить через билдер  Json { }
запровайдить эту зависимость через хилт и потом инжектить везде, где надо decodeFromString

JudjinGM replied:
Понял, поправлю!

avanisimov commented on app/src/main/java/app/cashadvisor/common/data/network/ErrorInterceptor.kt:
```diff
@@ -0,0 +1,51 @@
+            throw NetworkException.NoInternetConnection()
+        }
+        val response = with(chain) { proceed(request()) }
+        if (response.code !in (HttpURLConnection.HTTP_OK..HttpURLConnection.HTTP_CREATED)) {
+            throwErrorIfRequire(response.peekBody(Long.MAX_VALUE), response.code)
+        }
+        return response
+    }
```
а что будет, если throwErrorIfRequire не выбросит ошибку? как это обработает датасурс и репозиторий?

и еще
вместо response.peekBody(Long.MAX_VALUE) можно взять response.body() и проверить на null
peekBody создаст новый ResponseBody или выбросит Exception если body() null

JudjinGM replied:
я здесь не совсем понял по поводу response.body, если я вызову проверку на null, то так как я могу его получить только один раз, то response.body дальше не уйдет по цепочке. Или мне надо создать в конце новый response с тем же содержимым?

JudjinGM replied:
Поправил

avanisimov commented on app/src/main/java/app/cashadvisor/common/di/NetworkModule.kt:
```diff
@@ -0,0 +1,166 @@
+        return Retrofit.Builder()
+            .baseUrl(endpointBaseUrl)
+            .client(okHttpClient)
+            .addConverterFactory(Json.asConverterFactory("application/json".toMediaType()))
```
тут тоже надо инжектить подготовленный Json как объект-зависимость

JudjinGM replied:
Понял, поправлю!

avanisimov commented on app/src/main/java/app/cashadvisor/common/domain/Resource.kt:
```diff
@@ -0,0 +1,11 @@
+import app.cashadvisor.common.domain.model.ErrorEntity
+
+//пока назвал Resource, тк он может быть оберткой не только для данных из сети, а просто Result слишком популярное имя
+sealed class Resource<T>(val data: T? = null, val error: ErrorEntity? = null) {
+
+    class Success<T>(data: T) : Resource<T>(data)
+    class Error<T>(error: ErrorEntity) : Resource<T>(error = error)
+
+}
```
val data: T можно объявить прям в data class Success
тк для Error вообще не может быть data
```
sealed class Resource<T>() {

    data class Success<T>(val data: T) : Resource<T>()
    data class Error<T>(val error: ErrorEntity) : Resource<T>()

}
```

JudjinGM replied:
Понял, поправлю!

JudjinGM committed:
Refactoring with review suggestions #1

JudjinGM committed:
Implemented data/domain for register and login

JudjinGM committed:
Refactored:
1) Split ds, repo and interactor to login and register
2) Added state for confirmation code token in repos
3) Refactored hilt modules
3) Went through some refactoring (naming etc)

JudjinGM committed:
Fixed bug with login and register state

JudjinGM committed:
Small cleanup

JudjinGM committed:
added separate test fragment and viewmodel

JudjinGM committed:
small fix

JudjinGM committed:
Merge branch 'dev' into feature/55-data_domain_registration_login

# Conflicts:
#	app/build.gradle.kts
#	app/src/main/AndroidManifest.xml
#	app/src/main/java/app/cashadvisor/authorization/di/AuthorizationDataModule.kt
#	app/src/main/java/app/cashadvisor/authorization/domain/impl/AccountInformationInteractorImpl.kt
#	app/src/main/java/app/cashadvisor/authorization/presentation/ui/StartFragment.kt
#	app/src/main/java/app/cashadvisor/main/presentation/ui/MainActivity.kt
#	app/src/main/res/layout/fragment_start.xml
#	app/src/main/res/navigation/nav_graph.xml
#	settings.gradle.kts

JudjinGM committed:
Dev merge fix

JudjinGM committed:
Refactoring:
1) Changed names for mor accurate
2) Changed response and dto fields according backend changes

JudjinGM committed:
Merge branch 'dev' into feature/55-data_domain_registration_login

# Conflicts:
#	app/src/main/java/app/cashadvisor/main/presentation/ui/MainActivity.kt

JudjinGM committed:
some TestScreen improvements

JudjinGM committed:
Merge branch 'dev' into feature/55-data_domain_registration_login

# Conflicts:
#	app/src/main/java/app/cashadvisor/authorization/presentation/viewmodel/StartViewModel.kt

JudjinGM commented:
Проверить работоспособность можно в debug сборке, перейти в uiKit и выбрать Test fragment. 
Код подтверждения можно получить в свагере: https://212.233.78.3:8080/swagger/docs/index.html#/Dev/get_dev_confirmation_code_get , нажимайте Try it out и вводите email.

JudjinGM committed:
Fixed issue with response.peekBody and NPE possibility

MajoritySky2496 commented on app/src/main/java/app/cashadvisor/authorization/data/models/response/ConfirmRegisterResponse.kt:
```diff
@@ -0,0 +1,13 @@
+    @SerialName("status_code") val statusCode: Int,
+    @SerialName("access_token_life_time") val accessTokenLifeTime: Long,
+    @SerialName("refresh_token_life_time") val refreshTokenLifeTime: Long,
+    @SerialName("token_details") val tokenDetailsDto: TokenDetailsDto
```
Не увидел поля device_id, в свагере оно есть

JudjinGM replied:
Видимо недавно добавили, внесу изменения 👌

JudjinGM replied:
Добавил данное поле и для подтверждения регистрации и для подтверждения логина. Пока не понятно, что делать с этим полем, сейчас я его нигде не использую.

JudjinGM committed:
added device_id to register and login confirmation response

MajoritySky2496 commented on app/src/main/java/app/cashadvisor/authorization/data/impl/RegisterRemoteDataSourceImpl.kt:
```diff
@@ -0,0 +1,48 @@
+            val response = registerApiService.registerByEmail(
+                registerByEmailRequest = authenticationDataMapper.toRegisterByEmailRequest(inputDto)
+            )
+            authenticationDataMapper.toRegisterOutputDto(response)
```
Не понимаю для чего ты здесь получаешь  RegisterOutputDto?

JudjinGM replied:
Я здесь маплю в RegisterOutputDto, чтобы не таскать модель RegisterResponse, которая может иметь разные аннотации итд и она для работы с ретрофитом. Я это сделал, так как эти данные дальше в data слое уже не будут нужны и решил, что лучше замапить в модельку чисто data слоя.
